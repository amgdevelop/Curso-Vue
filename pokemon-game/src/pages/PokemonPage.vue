<template lang="">
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else :class="{green: correctAnswer === 1, red: correctAnswer === 2}">

    <h1>
        ¿Quién es este pokémon?
    </h1>

    <!--img-->
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <!--opciones-->
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />


    <template v-if="showAnswer">

      <h2 class="fade-in" >{{message}}</h2>
      <button @click="newGame" >
        Nuevo Juego
      </button>

    </template>
  </div>
</template>


<script>
import PokemonPicture from "../components/PokemonPicture.vue";
import PokemonOptions from "../components/PokemonOptions.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions.js";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
      correctAnswer: null,
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const rndInt = Math.floor(Math.random() * 4);

      this.pokemon = this.pokemonArr[rndInt];

      console.log(this.pokemon);
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;

      /* this.pokemon.id === selectedId
        ? (this.message = `Correcto, el pokemon es ${this.pokemon.name.toUpperCase()}`)
        : (this.message = `Lo siento, el pokemon es ${this.pokemon.name.toUpperCase()}`); */

      if (this.pokemon.id === selectedId) {
        this.message = `Correcto, el pokemon es ${this.pokemon.name.toUpperCase()}`;
        this.correctAnswer = 1;
      } else {
        this.message = `Lo siento, el pokemon es ${this.pokemon.name.toUpperCase()}`;
        this.correctAnswer = 2;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
      this.mixPokemonArray();
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>


<style scoped>
.red {
  background-color: red;
}
.green {
  background-color: green;
}
</style>