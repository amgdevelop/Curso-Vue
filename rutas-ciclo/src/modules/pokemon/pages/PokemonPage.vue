<template >
  <div>
    <h1>Pokemon # {{ id }}</h1>
    <div v-if="pokemon">
      <img :src="pokemon.sprites.front_default" :alt="pokemon.name" />
    </div>
  </div>
</template>


<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
  },

  data() {
    return {
      // id: this.id,
      pokemon: null,
    };
  },

  created() {
    this.getPokemon();
    /*  const { id } = this.$route.params;
    console.log(id);
    this.id = id; */
  },
  methods: {
    async getPokemon() {
      try {
        const pokemon = await fetch(
          `https://pokeapi.co/api/v2/pokemon/${this.id}`
        ).then((response) => response.json());

        this.pokemon = pokemon;
        console.log(pokemon);
      } catch (error) {
        this.$router.push("/");
      }
    },
  },
  watch: {
    id() {
      this.getPokemon();
    },
  },
};
</script>


<style scoped>
</style>