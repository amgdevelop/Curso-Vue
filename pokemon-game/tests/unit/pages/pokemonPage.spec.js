import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock'


describe('PokemonPage component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })


    test('debe llamar el mixPokemonArray en el mount', () => {

        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')

        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })


    test('debe hacer match con el snapshot cuando cargan los pokemon', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                };
            }
        })

        expect(wrapper.html()).toMatchSnapshot()



    })

    test('debe mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: true,
                    showAnswer: false,
                    message: "",
                };
            }
        })


        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        // pokemonPicture debe existir
        expect(picture.exists()).toBe(true)
        // pokemonOptcions debe existir
        expect(options.exists()).toBe(true)



        // pokemonPicture  attribute pokemonId === 5
        expect(picture.attributes('pokemonid')).toBe('1')

        //PokemonOptions attribute pokemons toBe true
        expect(options.attributes('pokemons')).toBeTruthy()
    })



    test('pruebas con checkAnswer', async () => {


        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                };
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Correcto, el pokemon es ${pokemons[0].name.toUpperCase()}`)


        await wrapper.vm.checkAnswer(10)
        console.log(wrapper.vm.message)
        expect(wrapper.vm.message).toBe(`Lo siento, el pokemon es ${pokemons[0].name.toUpperCase()}`)

    })




})