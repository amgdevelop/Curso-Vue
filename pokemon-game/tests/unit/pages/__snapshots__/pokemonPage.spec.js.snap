// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div>
  <h1> ¿Quién es este pokémon? </h1>
  <!--img-->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <!--opciones-->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
